Copilot/replace jszip with client zip

Arithmomaniac opened the pull request:
Addresses #37. Replaces JSZip with [client-zip](https://github.com/Touffy/client-zip) for improved performance and smaller bundle size.

In my smoke testing of "export audiobook", no re-zipping was done and the ZIP file download instantly.
MP3 download also worked.

This was largely written by Copilot, it's original description is below.

--------------

## Changes

- **CDN**: `jszip/3.7.1` → `client-zip/index.js` 
- **API migration**: JSZip's imperative API → client-zip's declarative file array
- **Folder handling**: Path-based naming (`"OEBPS/file.xhtml"`) instead of folder objects
- **Progress callbacks**: Removed (not supported by client-zip)

## Migration Pattern

```javascript
// Before (JSZip)
const zip = new JSZip();
zip.file("Part 001.mp3", blob);
let folder = zip.folder("metadata");
folder.file("cover.jpg", coverBlob);
const zipBlob = await zip.generateAsync({type: 'blob', compression: "STORE"});

// After (client-zip)
const files = [
  { name: "Part 001.mp3", input: blob },
  { name: "metadata/cover.jpg", input: coverBlob }
];
const zipBlob = await downloadZip(files).blob();
```

## Functions Updated

- `createMetadata()` - Returns file array instead of mutating zip object
- `createAndDownloadZip()` - Audiobook MP3 export
- `downloadEPUB()` - EPUB generation with META-INF and OEBPS structure
- `createContent()`, `makePackage()`, `makeToc()` - Helper functions for EPUB

EPUB spec requirements (mimetype first, uncompressed) maintained through file ordering.

Copilot committed:
Initial plan

Copilot committed:
Replace JSZip with client-zip library

Co-authored-by: Arithmomaniac <654897+Arithmomaniac@users.noreply.github.com>

Arithmomaniac committed:
use client-zip worker

PsychedelicPalimpsest commented on userscript.js:
```diff
@@ -12,7 +12,7 @@
 // @run-at        document-start
 // @icon          https://www.google.com/s2/favicons?sz=64&domain=libbyapp.com
-// @require       https://cdnjs.cloudflare.com/ajax/libs/jszip/3.7.1/jszip.min.js
+// @require       https://unpkg.com/client-zip@2.5.0/worker.js
```
I thought the url was `client-zip/index.js`? `worker.js` implies this is supposed to run in a webworker. I believe this is what causes the script to fail

Arithmomaniac replied:
index builds an esm module. worker builds an IIFE with downloadZip as a global export. The name "worker" is misleading.

https://github.com/Touffy/client-zip/blob/7fcc01b4b72f8bf0c8e9e26b5d29e5c6b22faab0/package.json#L14-L15

PsychedelicPalimpsest commented on userscript.js:
```diff
@@ -418,30 +425,28 @@
+        files.push(...downloadedFiles);
 
+        if (addMeta) {
+            const metadataFiles = await createMetadata();
```
What is the reason you are moving the createMetadata? When you do this it causes the download to take slightly longer

Arithmomaniac replied:
This should be fixed now. I need to smoke test it still

PsychedelicPalimpsest commented on userscript.js:
```diff
@@ -934,49 +948,53 @@
-        zip.file("mimetype", "application/epub+zip", {compression: "STORE"});
-        let metaInf = zip.folder("META-INF");
-        metaInf.file("container.xml", `<?xml version="1.0" encoding="UTF-8"?>
+        // Add META-INF files
```
How are you ensuring that the metadata is set to `STORE`? This is part of the EPUB spec, you must set it to store or some epub readers might fail

Arithmomaniac replied:
Compression isn't used at all in this library; it seems they explicitly keep that in mind when creating the file and central headers:

https://github.com/Touffy/client-zip/blob/7fcc01b4b72f8bf0c8e9e26b5d29e5c6b22faab0/src/zip.ts#L118-L128

https://github.com/Touffy/client-zip/blob/7fcc01b4b72f8bf0c8e9e26b5d29e5c6b22faab0/src/zip.ts#L165-L171

PsychedelicPalimpsest reviewed (commented):
The script does not run. I am not exactly sure why, but here are my idea:
1. You are using `worker.js`
2. Keep in mind this is an injected script, and therefore you MUST be careful with any WASM libs. 
3. Cors is a bitch

PsychedelicPalimpsest commented:
I'm sorry for the (extremely) late reply. I have completed a code review, it mostly looks fine, but the code does not run. 

I am alright with AI generated PRs, but I ask you ensure the code is running first. 

PS: I am unfamiliar with the `client-zip` library, but from my experience with FFmpeg wasm, I was forced to write a new frontend for it (see [ffmpeg.js](https://github.com/PsychedelicPalimpsest/FFmpeg-js)). So you cannot always just grab a script off a CDN and use it. 

**In addition if the script does not run fast enough it will error.** This is why I am running FFmpeg.js async by injecting the script into the header after the fact.

PsychedelicPalimpsest commented:
@Arithmomaniac

Arithmomaniac commented:
Yikes, I see I was out of my depth here re understanding this script.

I did run this locally (though only for audiobook export). I'll try and get to the bottom of this.

AI code reviews are a real pain to handle, I appreciate your forbearance 🙏

Arithmomaniac commented:
> The script does not run. I am not exactly sure why, but here are my idea:
> 
> 1. You are using `worker.js`
> 2. Keep in mind this is an injected script, and therefore you MUST be careful with any WASM libs.
> 3. Cors is a bitch

1. Explained inline.
2. client-zip This is wasm- and dependency-free.
3. Could be. What browser are you using? Can you detail the error you see?

Copilot committed:
Initial plan

Copilot committed:
Restore parallel execution for metadata creation

Move metadata creation to run concurrently with file downloads using Promise.all. This reduces total download time by fetching the cover image in parallel with MP3 files instead of sequentially.

Co-authored-by: Arithmomaniac <654897+Arithmomaniac@users.noreply.github.com>

Copilot committed:
Simplify by adding createMetadata to fetchPromises conditionally

As suggested, removed separate metadataPromise and downloadedFiles destructuring. Now just conditionally add createMetadata() to fetchPromises and flatten the results.

Co-authored-by: Arithmomaniac <654897+Arithmomaniac@users.noreply.github.com>

Copilot committed:
Revert "Simplify by adding createMetadata to fetchPromises conditionally"

This reverts commit 9e78ee0. Restoring the previous version with metadataPromise and array destructuring pattern.

Co-authored-by: Arithmomaniac <654897+Arithmomaniac@users.noreply.github.com>

Arithmomaniac committed:
Merge pull request #3 from Arithmomaniac/copilot/restore-parallel-execution-again

Restore parallel metadata fetching in audiobook export

PsychedelicPalimpsest commented:
Thanks for your quick action!
After some testing your script does work now. 

Interestingly, with the EPUB export function it does have a different result from the original script. 

**[Redacted examples due to copyright concerns]**

Where now formatting is being correctly inserted (where it was not before). This is likely being caused by this difference in the xhtml:

Before:
```html
<body lang="en-US" data-xml-lang="en-US" data-loc="8" id="undefined">
```
After:
```html
<body lang="en-US" data-xml-lang="en-US" data-loc="8" id="prl">
```

So congrats! Your PR fixed a bug I did not know existed.

PsychedelicPalimpsest commented:
Anyways, I will do a bit more testing, but I think your PR is ready to be merged.

Alexizk21 commented:
Hi, could someone develop this to be compatible with Voxabooks [https://voxabooks.com/]? I've managed to download manually using Widevine Proxy 2, but only chapter by chapter, which is very tedious with audiobooks that have many chapters. It would be great if someone could adapt and automate it for that site, please.

PsychedelicPalimpsest committed:
Update version number to 2025-11-18

PsychedelicPalimpsest commented:
Thank you for your contribution, it has now been merged!
